{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d2055b4f-28fe-4cf4-ac1b-b297ed6a5968",
   "metadata": {},
   "source": [
    "# Filtering variants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecdf69d2-141d-45c2-a098-d2cdc5560cb3",
   "metadata": {},
   "source": [
    "This is a [Jupyter notebook](https://jupyter.org/) containing a guide for filtering variant call (vcf) files that have been annotated using [SNPEff](https://pcingola.github.io/SnpEff/).\n",
    "\n",
    "This workflow was developed on output from the nfcore/sarek pipeline (https://nf-co.re/sarek) which is a variant calling and annotation pipeline built in the [Nextflow](https://www.nextflow.io/) workflow manager system. Sarek includes SNPEff annotation as part of the workflow. However, any SNPEff annotated vcf file can be used as input for this workflow, not just those generated by nfcore/sarek.\n",
    "\n",
    "This workflow was prepared by the [eResearch Office, QUT.](https://qutvirtual4.qut.edu.au/group/staff/governance/organisational-structure/academic-division/research-portfolio/research-infrastructure/eresearch)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d559f205-49b0-4de2-9f48-0e7119be322e",
   "metadata": {},
   "source": [
    "**********************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa0ee2ae-05f7-4d2d-be13-b22adb68e8e7",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Contents\n",
    "[1. How to use this notebook](#overview)\n",
    "\n",
    "[2. Requirements](#require)\n",
    "\n",
    "[3. Installing SNPEff](#install)\n",
    "\n",
    "[4. Filtration principles](#filt)\n",
    "\n",
    "[5. Filter by impact](#impact)\n",
    "\n",
    "[6. Look for overlaps](#overlap)\n",
    "\n",
    "[7. Merging vcfs](#merge)\n",
    "\n",
    "[7. Output as table](#table)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47679e0e-5308-4122-9ad6-4c39721808e9",
   "metadata": {},
   "source": [
    "***************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54f8b3c1-68ee-4e34-8a84-c0c98dd3b7e7",
   "metadata": {},
   "source": [
    "## How to use this notebook <a class=\"anchor\" id=\"overview\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73969ce6-0e39-4d79-a800-9f468c9cce6c",
   "metadata": {},
   "source": [
    "Juypter Notebooks run a 'kernel' that allow code to be run in code 'cells' in the Notebook. This Notebook is running the BASH kernel, which allows for commands to be run on QUTs HPC."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d89c273e-3629-4f15-a580-3652eb2307fb",
   "metadata": {},
   "source": [
    "You can run a code cell by clicking on the cell itself and clicking the run button (at the top of this Notebook), or by pressing shift+enter."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6008539f-3439-41a1-8568-63ad4872c194",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "As an example, run the following code cell to list the contents of your HPC home directory.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2a5d198-905d-47d0-941b-4ea4af1e4015",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ls $HOME"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "381d9617-ee8f-40a9-9774-09486c0ff7e5",
   "metadata": {},
   "source": [
    "Before every code cell is a colour-coded text box that tells you what the code cell will do. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67ed0e39-d2ee-4c7f-b3de-8237aa141f0d",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "A green text box indicates a code cell that must be run, without alteration, to complete the workflow.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31b55f29-8f46-46e6-bcef-d180ae643b26",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "A yellow text box indicates an optional code cell that doesn't have to be run to complete the workflow, but can be run to complete optional tasks.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b3a70aa-44cd-456a-829a-8ead827b9c4e",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "A blue text box indicates a code cell that requires user input - it must be run to complete the workflow and the user needs to modify the command in the cell. This is required for user-specific or project-specific operations, such as moving to a directory that contains the project data files.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37dc52fa-cf3e-423c-bbc7-3a2b3f7d03aa",
   "metadata": {},
   "source": [
    "In this guide you will find instructions and code cells to run that will filter vcf files that were annoated by SNPeff."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "034c9fde-68c9-42b5-bb97-4bf6dee3b727",
   "metadata": {},
   "source": [
    "*******************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f0ad50f-13bd-4113-920d-cb60ee745440",
   "metadata": {},
   "source": [
    "## Requirements <a class=\"anchor\" id=\"require\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddbc3f4b-487b-44b2-94b8-8c1bb6868c6f",
   "metadata": {},
   "source": [
    "You will need a QUT HPC account. If you are seeing this Notebook, you most likely already have a HPC account. Regardless, you can request an account be created, or request any other HPC or bioinformatics support, via the portal here: https://eresearchqut.atlassian.net/servicedesk/customer/portals"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edc88ab2-8f83-47df-8715-e945b7ea8843",
   "metadata": {},
   "source": [
    "You will also need to have all the annotated vcf files in a directory on the HPC. Sarek outputs these files as '..samplename..'snpEff.ann.vcf.gz. Depending on how you ran sarek or how your data is organised, these may be in separate directories based on treatment group or individual samples. \n",
    "\n",
    "It is recommended to copy all these files to a single directory. As directory and file structure varies per project, it's not possible to include the code to achieve this in this Notebook and thus the user must do this manually prior to running this workflow.\n",
    "\n",
    "A guide to copying and moving data via the Linux command line is [here](https://www.thegeekdiary.com/linux-command-line-basics-for-beginners-copy-and-move-files-and-directories-cp-and-mv-commands/), or you can contact the HPC staff to locate/move your data files via the portal [here](https://eresearchqut.atlassian.net/servicedesk/customer/portals)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d09587f6-9f7a-4852-b99d-6b43875b22d4",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "Enter the directory that contains your SNPEff annotated vcf files. You will need to change the directory path to where you moved your '..samplename..'snpEff.ann.vcf.gz files to. You can find this path by typing 'pwd' on the command-line when you are in that directory, or by contacting the HPC staff via the portal (above link). The structure of the below command should be `root_path=/directory/containing/my/vcf/files`.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e36807d1-87e8-493f-b03d-d86609260ddc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "root_path=/work/liver/nextflow/sarek/individual/sarek_VCFs_annotation/All_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db707631-00ef-4356-9b08-bf505cf7637a",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "Now move to the above directory (cd = change directory): \n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e6cd3f5-d1b0-467d-a676-ae6daf6a2f88",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cd $root_path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2235273d-8720-494f-9953-1821da0d2d6e",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "To see if you are in the correct directory, run the 'ls' code cell below. You should see a list of all your SNPEff annotated vcf files. If you don't see the files, you've entered the above location incorrectly and need to correct and re-run the above code cell.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d25418b8-c3cc-43bb-9185-1451f2115db1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1748e621-4df9-4896-9445-e5f4b720d7dc",
   "metadata": {},
   "source": [
    "****************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b643d635-28d2-4858-ae29-458c9b9516ec",
   "metadata": {},
   "source": [
    "## Installing SNPEff <a class=\"anchor\" id=\"install\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00d6a2f5-72f1-4f93-a99d-c4c97088b2b2",
   "metadata": {},
   "source": [
    "SNPEff needs to be installed first to run the various SNPEff command-line filtration options."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec8be7e2-e520-41ff-91f9-5b014c7e3f8b",
   "metadata": {
    "tags": []
   },
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "Create a new directory called 'snpeff' and download the latest version of SNPEff to that directory, then unzip the file.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a600b21-4fac-4e68-8f43-7b0a67b2ceb3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "mkdir snpeff\n",
    "cd snpeff\n",
    "wget https://snpeff.blob.core.windows.net/versions/snpEff_latest_core.zip\n",
    "unzip -q snpEff_latest_core.zip\n",
    "cd $root_path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "721962bf-74e6-4ea5-ac2c-33b52a8b565c",
   "metadata": {
    "tags": []
   },
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "SNPEff is based on Java, so the Java module will need to be loaded before running SNPEff.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63459706-0837-4635-83a6-a84cc8298aee",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "module load java"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04618d82-b8b8-497f-b817-2943a061371a",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "Check the version of SNPEff, to make sure it has been installed correctly\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "207904d9-d8eb-4b8e-bce1-de3a0442d1c7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "java -jar ./snpeff/snpEff/snpEff.jar -version"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d5a74bd-2f19-49b2-b11a-059052a7888f",
   "metadata": {},
   "source": [
    "********************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "893ca784-1c72-4927-b8a5-9501359caac1",
   "metadata": {},
   "source": [
    "## Filtration principles <a class=\"anchor\" id=\"filt\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93d9fd0f-637d-4a0d-8162-2279bc2a5728",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "If you look in the installed SNPEff directory (you can run the code cell below to do this), you'll see the two main tools: snpEff.jar and SnpSift.jar.\n",
    "    </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8de6473a-3701-475d-b8f9-c3bb1f567273",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ls ./snpeff/snpEff/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71e866fb-a93a-43bf-8e97-d38036e6ff78",
   "metadata": {},
   "source": [
    "snpEff.jar\n",
    "> is a variant annotation and effect prediction tool. It annotates and predicts the effects of genetic variants (such as amino acid change)\n",
    "\n",
    "SnpSift.jar\n",
    ">is a toolbox that allows you to filter and manipulate annotated files."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a24ec1e5-2724-47e7-b15b-d7e67dfbe1e0",
   "metadata": {},
   "source": [
    "snpEff has already been run on your samples, as part of the nfcore/sarek workflow, creating the '..samplename..'snpEff.ann.vcf.gz files we're working with here. \n",
    "\n",
    "In this Notebook we will be using SnpSift.jar to filter your samples by certain parameters, such as high impact variants.\n",
    "\n",
    "SnpSift.jar is highly flexible and can be used to filter by quality, read depth, genotype and any of the annotation fields generated by snpEff.jar (such as frameshift mutations, synonymous, missense, splice regions, intron, exon, etc, etc).\n",
    "\n",
    "A list and explanation of the filtration options can be seen here:\n",
    "\n",
    "https://pcingola.github.io/SnpEff/ss_filter/\n",
    "\n",
    "And a list of the possible annotation fields can be seen here:\n",
    "\n",
    "https://pcingola.github.io/SnpEff/adds/VCFannotationformat_v1.0.pdf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "795c6f9a-49bc-4362-8c92-440d51caca13",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "By following this guide, you have installed SnpSift.jar in ./snpeff/snpEff/SnpSift.jar and can run any of the examples in the link above by adding './snpeff/snpEff/' before 'SnpSift.jar' in any command.\n",
    "\n",
    "For example, if you wanted to run the first example in https://pcingola.github.io/SnpEff/ss_filter/ (filter out samples with quality less than 30), you could run the following (change 'variants.vcf' to one of your variant files and 'filtered.vcf' to a suitable name for the output file):\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "518b815d-30c7-4718-8f35-d5341664f12f",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat variants.vcf | java -jar ./snpeff/snpEff/SnpSift.jar filter \" ( QUAL >= 30 )\" > filtered.vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9847a9bf-eba3-40be-a48b-90ff044d5fc4",
   "metadata": {},
   "source": [
    "SnpSift.jar can do more than filter samples, it can extract specific fields from your annotated vcf file, join by genomic regions, split by chomosome, etc. See all the SnpSift.jar commands here:  https://pcingola.github.io/SnpEff/ss_introduction/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5943ca5-83fc-4ea1-bc4a-a3428c064952",
   "metadata": {},
   "source": [
    "***************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b82d3af-60ab-4896-aa39-0c45440ad50b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Filter by impact <a class=\"anchor\" id=\"impact\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7572e9ff-b0ad-4851-8ecd-a981a9b78761",
   "metadata": {},
   "source": [
    "SNPEff annotates variants by HIGH, MODERATE, LOW or MODIFIER impact.\n",
    "\n",
    "https://pcingola.github.io/SnpEff/se_inputoutput/#impact-prediction\n",
    "\n",
    ">HIGH\tThe variant is assumed to have high (disruptive) impact in the protein, probably causing protein truncation, loss of function or triggering nonsense mediated decay.\n",
    "\n",
    ">MODERATE\tA non-disruptive variant that might change protein effectiveness.\n",
    "\n",
    ">LOW\tAssumed to be mostly harmless or unlikely to change protein behavior.\n",
    "\n",
    ">MODIFIER\tUsually non-coding variants or variants affecting non-coding genes, where predictions are difficult or there is no evidence of impact."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4eeae756-7763-404b-a186-b2c5e536cb6e",
   "metadata": {},
   "source": [
    "A list of all the effects (e.g. deletions, duplications, frame shift, etc) can be seen here, including if their impact category: https://pcingola.github.io/SnpEff/se_inputoutput/#effect-prediction-details"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca2f2021-3bd7-4ebb-b003-3da2298d2c0a",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "Create a main directory to store filtered results from each sample group, which will be in their own sample group subdirectories:\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19950dca-fd25-4302-a061-5b109751b2d4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cd $root_path\n",
    "mkdir Filtered"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fc2ffbf-2014-4f12-a5f1-aaf7eab39dc4",
   "metadata": {},
   "source": [
    "You can filter out variants by impact (e.g. only keep high impact variants)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ee47e8c-7716-4b6a-b8ea-4b307646e739",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "Run one or more of the three following code cells. The first cell filters out HIGH impact variants, the second cell MODERATE and the third cell LOW. You may only be interested in HIGH impact variants, in which case just run the first cell. Note: this will not alter the original vcf files. New vcf files will be created in 'filtered_high', 'filtered_moderate' or 'filtered_low' directories. These commands may take several minutes to run.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc8d6be0-efcb-4d44-bb3f-c2c32d4711f7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "mkdir Filtered/filtered_high\n",
    "for f1 in *.vcf*\n",
    "do\n",
    "echo filtering $f1\n",
    "java -jar ./snpeff/snpEff/SnpSift.jar filter \"ANN[*].IMPACT has 'HIGH'\" $f1 > \"Filtered/filtered_high/high_$f1\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "469e5bc9-d001-4c27-9622-0d19c3a85718",
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir Filtered/filtered_moderate\n",
    "for f1 in *.vcf*\n",
    "do\n",
    "echo filtering $f1\n",
    "java -jar ./snpeff/snpEff/SnpSift.jar filter \"ANN[*].IMPACT has 'MODERATE'\" $f1 > \"Filtered/filtered_moderate/moderate_$f1\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcb8a8cb-2a27-48ab-b2cf-85b2a2dae2d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir Filtered/filtered_low\n",
    "for f1 in *.vcf*\n",
    "do\n",
    "echo filtering $f1\n",
    "java -jar ./snpeff/snpEff/SnpSift.jar filter \"ANN[*].IMPACT has 'LOW'\" $f1 > \"Filtered/filtered_low/low_$f1\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3925d62a-c2e3-45aa-b1f0-85839724a372",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "To check that the filtered vcf files were created by listing the files in the 'filtered_high' directory. You can check the MODERATE of LOW results also by changing 'filtered_high' to 'filtered_moderate' or 'filtered_low' in the code cell below.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58d23bee-6424-4427-8689-9ddb89a352ea",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ls Filtered/filtered_high"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4626af32-4da8-42bc-8dca-4924bd4a876d",
   "metadata": {},
   "source": [
    "*******************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd32f5d7-eceb-401c-89b7-a88df1daeb28",
   "metadata": {},
   "source": [
    "## Look for overlaps <a class=\"anchor\" id=\"overlap\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3d1c807-fb77-4690-9b8a-409a0ce8a748",
   "metadata": {},
   "source": [
    "Usually you will have multiple samples per treatment group, which means you'll want to find common variants across multiple samples. This is done by creating intersections between two or more vcf files."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "960c82c4-d2d2-4e23-af5f-1c7d09aa719b",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "We'll be using bcftools to look for intersects. bcftools is already installed on the HPC as a module. Load the bcftools module:\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3314d64c-d88c-4734-9748-403b4c0b56ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "module load bcftools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96ef491f-2bb7-4819-8bda-766a7aec8726",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "Create a main directory to store each sample group intersect results, which will be in their own sample group subdirectories:\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "071161ee-ad81-4726-bf58-5f911850ffe5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cd $root_path\n",
    "mkdir Intersect"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8ac527c-e7b5-4c4d-bc95-6ce7a1af008e",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "Now select a target treatment group.\n",
    "In the code cell below, enter a character string that is unique to the target treatment group, i.e. is used in the name of every sample of only that group (i.e. change 'group=...' to the target group name your sample names).\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "322eb94a-c7b8-4287-8391-14661840658d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "group=ALD_EtOH"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a789775d-8ed4-4d6d-ab96-81dd503caaa6",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "If you can't remember your sample names, run 'ls' to see them:\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc483ba3-9e32-4fc5-a9a2-a69779d66249",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16555e81-da1a-48e4-af91-54bcc30c7e1c",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "You can also check that you are only pulling out the correct samples. The samples you will be intersecting are: \n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "077db93c-5988-476c-8bcb-695ff1743a48",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ls *$group*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "963d68a4-232f-443f-84ed-9ebc288e644a",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "Bcftools requires that the vcf file is compressed and indexed. First we'll compress the vcf files:\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a98d1e8-625c-4eb6-8242-4d7b1069b3a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir Intersect/$group\"_intersect\"\n",
    "for f1 in *$group*\n",
    "do\n",
    "echo compressing $f1\n",
    "bgzip -c $f1 > Intersect/$group\"_intersect\"/$f1\".gz\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df29e4c9-be30-4bf7-b412-1ceb5c744945",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "Then we'll index the vcf files:\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e6ee17b-cec8-40f6-916d-c79d9dc36bcc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cd Intersect/$group\"_intersect\"\n",
    "for f1 in *$group*\".gz\"\n",
    "do\n",
    "echo indexing $f1\n",
    "bcftools index $f1\n",
    "done\n",
    "cd $root_path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e52547fb-b532-49f1-a9b2-ccfb9b6b9462",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "Now we can check for intersections between these samples:\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d2ccafb-5c18-470d-a172-7dc359da0da9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cd Intersect/$group\"_intersect\"\n",
    "bcftools isec -n+7 -c all *$group*\".gz\" > $group\"_sharedPositions.txt\"\n",
    "cd $root_path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e44b5a2a-cda6-41fd-8b43-1af70e7eb921",
   "metadata": {},
   "source": [
    "**Note: the '-n+7' option says to output variants that intersect in 7 or more of your samples. You should adjust this to a number suitable to the total number of samples in the treatment group. For example, if you have n=20, setting -n+10 would output variants that are in 50% of your samples. -n+20 would only output variants that are in all samples, -n+2 in 2 or more samples, etc.**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a9e15dd-1951-4ae6-a4b4-5deecf695518",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "Find the number of samples in your treatment group:\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1541cdeb-4428-4c53-bac8-9fc130f7a3cd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ls Intersect/$group\"_intersect\"/*$group*\".gz\" | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc5b4015-157c-426d-8502-79c339216545",
   "metadata": {},
   "source": [
    "Now that you have a table of intersecting variants, you need to pull out just these intersects from your vcf files. For this we use bcftools view:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d192054-363e-490d-9152-5d407fafa1bb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cd $root_path\n",
    "cd Intersect/$group\"_intersect\"\n",
    "for f1 in *$group*\".gz\"\n",
    "do\n",
    "echo filtering $f1\n",
    "bcftools view -T $group\"_sharedPositions.txt\" $f1 -Oz > \"sharedPositions_\"$f1\n",
    "done\n",
    "cd $root_path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8112229f-4121-41f8-81c9-314996383ff4",
   "metadata": {},
   "source": [
    "These filtered vcfs need to be re-indexed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41b57bad-e4ac-42b1-aadb-99f248851a02",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cd Intersect/$group\"_intersect\"\n",
    "for f1 in sharedPositions*.vcf.gz\n",
    "do\n",
    "echo indexing $f1\n",
    "bcftools index $f1\n",
    "done\n",
    "cd $root_path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "660a1437-7c22-49bc-aaea-173fd0605c08",
   "metadata": {},
   "source": [
    "You can find intersects in other groups by going to the treatment group selection code ('group=..'), changing the group name and then running through the process again. You can do this for each treatment group. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78be955f-5588-4489-9c3e-01894ffec505",
   "metadata": {},
   "source": [
    "*******************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58ae0428-5c42-4c3e-b869-1e4e66b29c3f",
   "metadata": {},
   "source": [
    "## Merging vcfs <a class=\"anchor\" id=\"merge\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81e93ad9-e49f-4e12-82f0-243c0c1c58e8",
   "metadata": {},
   "source": [
    "Once you have filtered and/or found intersecting variants between samples, you can merge all treatment group samples into a single vcf file. We'll be using bcftools to accomplish this. If you've started a new session, make sure you run the 'module load bcftools' code cell in the previous section."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2b26b89-5f5d-49b4-b4d3-1cacd5c3d80c",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "Create a main directory to store merged vcf files for each sample group, which will be in their own sample group subdirectories:\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a6cb11d-0198-4963-8128-6a1de62d85b6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cd $root_path\n",
    "mkdir Merged"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43e8f15d-9cd3-4064-84fc-f5e3c48be616",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "Enter a subdirectory that contains your intersected results\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74a095d3-c24e-4154-b6a9-b0ddac67d601",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "intdir=ALD_EtOH_intersect"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56c2577c-9afc-44c0-97d3-0c056f1e61ee",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "See here for a list of the intersected samples group directories:\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89da2c4b-4c4f-46cb-a94d-eb13fdd81502",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cd $root_path\n",
    "ls -d Intersect/*/ | cut -f2 -d'/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "556ac5a9-ec57-4250-9e2a-978f303415e0",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "Use bcftools merge to merge the vcf files:\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d0b9006-b8a1-477d-b4f0-c1811a3b6270",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "bcftools merge Intersect/$intdir/sharedPositions*.vcf.gz > Merged/$intdir\"merged.vcf\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4a906c3-9a11-42f6-bc4b-1787aaf3244f",
   "metadata": {},
   "source": [
    "*************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72dc7a4e-43e0-40b7-b540-647c1e7eac78",
   "metadata": {},
   "source": [
    "## Output as table <a class=\"anchor\" id=\"table\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "620c1a28-ca03-4de0-88a7-9b2d0ad9ecad",
   "metadata": {},
   "source": [
    "SnpSift.jar can extract any fields in a vcf, including any annotation fields, outputting this as a table. This is useful for importing your filtered results into another program like R or Excel.\n",
    "\n",
    "https://pcingola.github.io/SnpEff/ss_extractfields/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08c63a74-a2f4-40e4-8093-132baaaa4e3f",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "For example, if you wanted to output chromosome, position, ID and allele frequency from a vcf file, you could run the following (change 'variants.vcf' to one of your vcf files, and 'table.txt' to a more informative name):\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfe884bf-a12c-4ae6-8c52-8b4c5f9e17aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "java -jar ./snpeff/snpEff/SnpSift.jar extractFields variants.vcf CHROM POS ID AF > table.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8651d078-70b0-4b6a-af76-372a21b7589c",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "Then you can either download the table or view it like so (change 'table.txt' to the same output filename that you used above)\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6aba01f-7088-446c-ab77-80ddd480b0ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat table.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8126c8cb-7b7a-46c7-aa48-7c74cbd83b25",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9ccef52-4874-4e3b-9552-46e7623e716a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffdb3e93-6b95-4e91-a59e-71edcdbfe7c9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c1f3a9b-b95c-4b52-bf5e-2e4719461d43",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cb4dbfd-328e-4fae-952b-117ddc3fc2a2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d140fa2-274b-4261-b5ee-1e30240d87eb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash .conda-jupyterhub *",
   "language": "bash",
   "name": "conda-env-.conda-jupyterhub-bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
